# 🧠 Machine Learning Research Roadmap

## Day 2: The Machine Learning Research Workflow

> From Research Questions to Evidence-Based Conclusions

---

📦 **Dataset Used**

Titanic Research Starter Pack

---

📚 **Companion GitHub Chapter**

(https://github.com/Yashashree-Joshi/machine-learning-research-roadmap/tree/5597bd1c90a40af88b060f07169eb18e5d1b43f9/Day-02-The-ML-Research-Workflow)

---

# 🤔 A Common Misconception

Many beginners think Machine Learning Research looks like this:

```text
Dataset
   ↓
Model
   ↓
Accuracy
```

Build a model.

Measure accuracy.

Move on.

---

Researchers think differently.

They ask:

- What problem am I solving?
- What question am I investigating?
- What hypothesis am I testing?
- How will I evaluate the results?

Research is not random experimentation.

Research is a structured process for generating evidence.

In this chapter, we'll learn the complete Machine Learning Research Workflow and apply it to a real-world example using the Titanic dataset.

In [1]:
#Libraries needed
import pandas as pd

In [2]:
train = pd.read_csv(
    "/kaggle/input/datasets/yashashree01/titanic-dataset-for-ml-research-roadmap/train.csv"
)

print("Dataset Shape:", train.shape)

train.head()

Dataset Shape: (891, 12)


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [3]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
train.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


# 📊 Understanding the Dataset

Before formulating a research question, researchers first understand the available data.

From the outputs above, we can observe:

- The dataset contains passenger demographic and travel information.
- Some features contain missing values.
- The target variable is **Survived**.
- Features have different data types and scales.

---

### Initial Observations

Some questions immediately arise:

1. Do missing values affect model performance?
2. Which features are most important for predicting survival?
3. Does feature selection improve model accuracy?
4. Which evaluation metric should we prioritize?

These observations help transform a dataset into a research opportunity.

# ❓ Formulating a Research Question

A dataset alone is not research.

Research begins when we ask a specific question.

Based on our observations, we will investigate:

> Does feature selection improve Titanic survival prediction performance?

---

## Why This Question?

We observed:

- Multiple input features
- Missing values
- Different feature types

Not every feature may contribute equally to prediction.

Removing less informative features may improve model performance and reduce noise.

This gives us a testable research question.

# 📝 Hypothesis

A hypothesis is a testable prediction.

For our research question:

> Does feature selection improve Titanic survival prediction performance?

We propose the following hypothesis:

### H₁ (Alternative Hypothesis)

Removing less informative features will improve model performance by reducing noise and helping the model focus on relevant information.

### H₀ (Null Hypothesis)

Feature selection has no significant effect on model performance.

---

The next step is to design an experiment that can test this hypothesis.

In [5]:
# Check missing values

train.isnull().sum().sort_values(ascending=False)

Cabin          687
Age            177
Embarked         2
PassengerId      0
Name             0
Pclass           0
Survived         0
Sex              0
Parch            0
SibSp            0
Fare             0
Ticket           0
dtype: int64

# ⚙️ Experiment Design

To test our hypothesis, we will build a baseline model using a selected set of features.

Future chapters will compare this model against alternative feature-selection strategies.

---

## Evaluation Metric

We will use:

- Accuracy
- F1 Score


In [6]:
missing = train.isnull().sum().sort_values(ascending=False)

missing[missing > 0]

Cabin       687
Age         177
Embarked      2
dtype: int64

# 🔍 Interpreting the Evidence

The missing-value analysis reveals three important observations:

| Feature | Missing Values |
|----------|---------------|
| Cabin | 687 |
| Age | 177 |
| Embarked | 2 |

---

### Research Decisions

#### Cabin

More than 75% of the values are missing.

Decision:

Drop the feature.

---

#### Age

The feature is potentially important for survival prediction.

Decision:

Keep the feature and impute missing values.

---

#### Embarked

Only two values are missing.

Decision:

Fill missing values using the most frequent category.

---

This process demonstrates an important research principle:

> Data-driven decisions should be justified using evidence.

In [7]:
# Select features for our baseline model

features = [
    "Pclass",
    "Sex",
    "Age",
    "SibSp",
    "Parch",
    "Fare",
    "Embarked"
]

train_model = train[features + ["Survived"]].copy()

train_model.head()

,Pclass,Sex,Age,SibSp,Parch,Fare,Embarked,Survived
0,3,male,22.0,1,0,7.2500,S,0
1,1,female,38.0,1,0,71.2833,C,1
2,3,female,26.0,0,0,7.9250,S,1
3,1,female,35.0,1,0,53.1000,S,1
4,3,male,35.0,0,0,8.0500,S,0


In [8]:
# Handle missing values

train_model["Age"] = train_model["Age"].fillna(
    train_model["Age"].median()
)

train_model["Embarked"] = train_model["Embarked"].fillna(
    train_model["Embarked"].mode()[0]
)

# Convert categorical features

train_model["Sex"] = train_model["Sex"].map(
    {"male": 0, "female": 1}
)

train_model = pd.get_dummies(
    train_model,
    columns=["Embarked"],
    drop_first=True
)

train_model.head()

,Pclass,Sex,Age,SibSp,Parch,Fare,Survived,Embarked_Q,Embarked_S
0,3,0,22.0,1,0,7.2500,0,False,True
1,1,1,38.0,1,0,71.2833,1,False,False
2,3,1,26.0,0,0,7.9250,1,False,True
3,1,1,35.0,1,0,53.1000,1,False,True
4,3,0,35.0,0,0,8.0500,0,False,True


# 🤖 Baseline Model

We now train our first machine learning model.

This serves as a baseline against which future experiments can be compared.

For simplicity, we will use Logistic Regression.

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score

# Features and target

X = train_model.drop("Survived", axis=1)
y = train_model["Survived"]

# Train-test split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# Train model

model = LogisticRegression(max_iter=1000)

model.fit(X_train, y_train)

# Predictions

y_pred = model.predict(X_test)

# Metrics

accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Accuracy:", round(accuracy, 4))
print("F1 Score:", round(f1, 4))

Accuracy: 0.8101
F1 Score: 0.7639


# 📈 Evaluation

Our baseline Logistic Regression model achieved:

| Metric | Score |
|----------|----------|
| Accuracy | 0.8101 |
| F1 Score | 0.7639 |

---

## What Do These Results Mean?

The model correctly predicted survival outcomes for approximately 81% of passengers in the test set.

The F1 Score of 0.7639 indicates a good balance between precision and recall.

---

## Research Perspective

The goal of this experiment was not to achieve the highest possible score.

The goal was to establish a baseline model that future experiments can improve upon.

Researchers rarely begin with the best model.

They begin with a reproducible baseline and then test hypotheses through systematic experimentation.

# 🔍 Conclusion

In this chapter, we followed the complete Machine Learning Research Workflow:

```text
Problem
↓
Research Question
↓
Hypothesis
↓
Dataset Exploration
↓
Evidence Gathering
↓
Experiment
↓
Evaluation
```

Using the Titanic dataset, we:

- Formulated a research question
- Created a hypothesis
- Investigated missing values
- Selected features
- Built a baseline model
- Evaluated results

This demonstrates how Machine Learning Research moves from questions to evidence-based conclusions.

---

### Key Lesson

Research is not:

Dataset → Model → Accuracy

Research is:

Question → Hypothesis → Experiment → Evidence → Conclusion

# 🔗 Continue Learning

This notebook is part of the **Machine Learning Research Roadmap** series.

### GitHub Repository

https://github.com/Yashashree-Joshi/machine-learning-research-roadmap

### Day 2 GitHub Chapter

https://github.com/Yashashree-Joshi/machine-learning-research-roadmap/tree/main/Day-02-The-ML-Research-Workflow

---

### Resources

📦 Titanic Research Starter Pack

(https://www.kaggle.com/datasets/yashashree01/titanic-dataset-for-ml-research-roadmap)

---

### Next Chapter

🚀 Day 3 — Research Question Formulation

Learn how researchers transform problems into strong, testable research questions.